src/sentry/snuba/metrics.py

```diff
@@ -0,0 +1,330 @@
+import itertools
+import math
+import random
+import re
+from datetime import datetime, timedelta
+from typing import Dict, List, Tuple
+
+import pytz
+
+from sentry.models import Project
+from sentry.snuba.sessions_v2 import (  # TODO: unite metrics and sessions_v2
+    InvalidField,
+    InvalidParams,
+    get_date_range_from_params,
+    parse_stats_period,
+    to_datetime,
+    to_timestamp,
+)
+
+FIELD_REGEX = re.compile(r"^(\w+)\(((\w|\.|_)+)\)$")
+TAG_REGEX = re.compile(r"^(\w|\.|_)+$")
+
+OPERATIONS = (
+    "avg",
+    "count_unique",
+    "count",
+    "max",
+    "p50",
+    "p75",
+    "p90",
+    "p95",
+    "p99",
+    "sum",
+)
+
+
+def parse_field(field: str) -> Tuple[str, str]:
+    matches = FIELD_REGEX.match(field)
+    try:
+        operation = matches[1]
+        metric_name = matches[2]
+    except (IndexError, TypeError):
+        raise InvalidField(f"Failed to parse '{field}'. Must be something like 'sum(my_metric)'.")
+    else:
+        if operation not in OPERATIONS:
+
+            raise InvalidField(
+                f"Invalid operation '{operation}'. Must be one of {', '.join(OPERATIONS)}"
+            )
+
+        return operation, metric_name
+
+
+def verify_tag_name(name: str) -> str:
+
+    if not TAG_REGEX.match(name):
+        raise InvalidParams(f"Invalid tag name: '{name}'")
+
+    return name
+
+
+def parse_tag(tag_string: str) -> dict:
+    try:
+        name, value = tag_string.split(":")
+    except ValueError:
+        raise InvalidParams(f"Expected something like 'foo:\"bar\"' for tag, got '{tag_string}'")
+
+    return (verify_tag_name(name), value.strip('"'))
+
+
+def parse_query(query_string: str) -> dict:
+    return {
+        "or": [
+            {"and": [parse_tag(and_part) for and_part in or_part.split(" and ")]}
+            for or_part in query_string.split(" or ")
+        ]
+    }
+
+
+MAX_POINTS = 1000
+ONE_DAY = timedelta(days=1).total_seconds()
+ONE_HOUR = timedelta(hours=1).total_seconds()
+ONE_MINUTE = timedelta(minutes=1).total_seconds()
+
+
+def get_constrained_date_range(
```

Comment:
Should we import this from sessions_v2, or does this implementation have material differences?

Reply:
It had fewer constraints, but you're right - I solved it with an import now.